File: 3794-find-the-minimum-amount-of-time-to-brew-potions/3794-find-the-minimum-amount-of-time-to-brew-potions.java
class Solution {
    public long minTime(int[] skill, int[] mana) {
        int n = skill.length;
        int m = mana.length;
        long[] finishtime = new long[n];//(n, 0);
        for (int j = 0; j < m; ++j) {
            finishtime[0] += (long) mana[j] * skill[0];

            for (int i = 1; i < n; ++i) {
                finishtime[i] = Math.max(finishtime[i],finishtime[i - 1]) + (long) mana[j] * skill[i];
            }

            for (int i = n - 1; i > 0; --i) {
                finishtime[i - 1] = finishtime[i] - (long) mana[j] * skill[i];
            }
        }
        return finishtime[n - 1];
    }
}

File: 3794-find-the-minimum-amount-of-time-to-brew-potions/3794-find-the-minimum-amount-of-time-to-brew-potions.cpp
class Solution {
public:
    long long minTime(vector<int>& skill, vector<int>& mana) {
        int n = skill.size();
        int m = mana.size();
        vector<long long> finishtime(n, 0);
        for (int j = 0; j < m; ++j) {
            finishtime[0] += 1LL* mana[j] * skill[0];

            for (int i = 1; i < n; ++i) {
                finishtime[i] = max(finishtime[i],finishtime[i - 1]) + 1LL * mana[j] * skill[i];
            }

            for (int i = n - 1; i > 0; --i) {
                finishtime[i - 1] = finishtime[i] - 1LL* mana[j] * skill[i];
            }
        }
        return finishtime[n - 1];
    }
};